### PR TITLE
Introduce service levels

### PR DESCRIPTION
**Pull request by psarna**
This series introduces service level syntax borrowed from https://docs.scylladb.com/using-scylla/workload-prioritization/ , but without workload prioritization itself - just for the sake of using identical syntax to provide different parameters later. The new parameters may include:
 * per-service-level timeouts
 * oltp/olap declaration, which may change the way Scylla treats long requests - e.g. time them out (the oltp way) or keep them sustained with empty pages (the olap way) 

Refs #7617

### CODE REVIEW

**Comment by psarna** [comment]
The tests run fine, but actually running Scylla fails with current code, because I missed one bit - setting a proper "data accessor" for the service level controller during initialization. Will fix tomorrow.

**Comment by avikivity** [comment]
@eliransin please review.

It's not a backport because this is upstream (strictly speaking it's a forward port). Please phrase it as a regular new feature introduction.

I wonder if we should use `CREATE SERVICE LEVEL` instead of `CREATE SERVICE_LEVEL` (enterprise can support the old syntax for compatibility).

**Comment by psarna** [comment]
> @eliransin please review.
> 
> It's not a backport because this is upstream (strictly speaking it's a forward port). Please phrase it as a regular new feature introduction.
> 
> I wonder if we should use `CREATE SERVICE LEVEL` instead of `CREATE SERVICE_LEVEL` (enterprise can support the old syntax for compatibility).

Yes please. I thought the syntax is already settled due to https://docs.scylladb.com/using-scylla/workload-prioritization/ , but if it's not, I'm for it.

**Comment by psarna** [comment]
It will actually be best for the backporting purposes (actual backporting) if both `SERVICE_LEVEL` and `SERVICE LEVEL` syntax is accepted. I'll add a separate patch that adds the alternative syntax for reviewing purposes and if everyone likes it, it can stay.

**Comment by psarna** [comment]
I'm also halfway through porting per-role timeouts to per-service-level timeouts here: https://github.com/psarna/scylla/commits/add_per_role_timeouts_5 . The branch is temporarily rebased on top of this pull request, since it heavily depends on service levels. The translating effort is still not finished, but I should be done with it next week.

**Comment by avikivity** [comment]
> It will actually be best for the backporting purposes (actual backporting) if both `SERVICE_LEVEL` and `SERVICE LEVEL` syntax is accepted. I'll add a separate patch that adds the alternative syntax for reviewing purposes and if everyone likes it, it can stay.

It's enough to support just `SERVICE LEVEL`. The grammar can add "_?" to K_SERVICE_LEVEL for compatibility.

**Comment by psarna** [comment]
rebased on top of current master to keep the pull request fresh and crusty

**Comment by avikivity** [comment]
@eliransin review ping

**Comment by psarna** [comment]
monthly rebase, conflicts resolved

**Comment by avikivity** [comment]
@eliransin re-ping

**Comment by psarna** [comment]
monthly rebase, conflicts resolved

**Comment by avikivity** [comment]
@slivne please help

File: service/qos/qos_common.cc [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "qos_common.hh"
+namespace qos {
+
+bool operator==(const service_level_options &lhs, const service_level_options &rhs) {
+    return &lhs < &rhs;
```
Comment (eliransin) [inline service/qos/qos_common.cc:26]: It looks wrong. I guess you meant `==` instead of `<` ?

File: service/priority_manager.hh [hunk 2]
```diff
@@ -24,6 +24,7 @@
 
 #include "seastarx.hh"
 
+
```
Comment (eliransin) [inline service/priority_manager.hh:27]: stray new line

File: auth/standard_role_manager.hh [hunk 3]
```diff
@@ -98,5 +106,4 @@ private:
 
     future<> modify_membership(std::string_view role_name, std::string_view grantee_name, membership_change) const;
 };
-
```
Comment (eliransin) [inline auth/standard_role_manager.hh:101]: stray remove line?

**Review by eliransin — COMMENTED** [review]
My final question is, did you test that queries are now service level aware (even with a print).
And a final note:
I am almost sure it will break upgrades to enterprise. It should be pretty easy to test. If it indeed breaks the service_level mechanism we should open an issue (int enterprise repo) to fix this.

Beside my minor comments - LGTM, sorry for the very long delay :slightly_smiling_face:

File: service/qos/qos_common.cc (hunk 1 above)
Comment (psarna) [reply service/qos/qos_common.cc:26]: yup, definitely

Comment (psarna) [reply service/qos/qos_common.cc:26]: Also, this comparison operator is actually unnecessary in open-source (it's already removed in a series that sits on top of this one: 8f4d1a8d40bdbfd820b53429cd53a2a5dbcb71ca ), so I'll actually rework this port to get rid of this operator entirely.

**Comment by psarna** [comment]
Right, updates to enterprise would get a little tricky - they'd have to modify the schemas for internal tables which keep service level information.

**Comment by eliransin** [comment]
> Right, updates to enterprise would get a little tricky - they'd have to modify the schemas for internal tables which keep service level information.

I think I might have an Idea about how to solve this, I will send an RFC.

**Comment by psarna** [comment]
I applied the review comment, but before proceeding I'll still try to also rebase https://github.com/scylladb/scylla/pull/7617 on top of this one first. Otherwise, it's impossible to test if the service level infrastructure was ported correctly, because currently the service levels have 0 parameters.

File: docs/service_levels.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+```CREATE TABLE system_distributed.service_levels (
+    service_level text PRIMARY KEY);
+```
+
+The table is used to store and distribute the service levels configuration.
+The table column names meanings are:
+*service_level* - the name of the service level.
+
+This table is currently a stub and does nto hold any parameters yet.
```
Comment (avikivity) [inline docs/service_levels.md:37]: s/nto/not/

File: cql3/Cql.g [hunk 5]
```diff
@@ -1767,6 +1875,12 @@ basic_unreserved_keyword returns [sstring str]
         | K_LIKE
         | K_PER
         | K_PARTITION
+        | K_SERVICE_LEVEL
+        | K_ATTACH
+        | K_DETACH
+        | K_SERVICE_LEVELS
+                | K_ATTACHED
+                | K_FOR
         | K_GROUP
```
Comment (avikivity) [inline cql3/Cql.g:1892]: Is the extra indent intended?

File: service/qos/service_level_controller.hh [hunk 6]
```diff
@@ -222,7 +222,7 @@ private:
      */
     service_level& get_service_level(sstring service_level_name) {
         auto sl_it = _service_levels_db.find(service_level_name);
-        if (sl_it == _service_levels_db.end() && sl_it->second.marked_for_deletion) {
+        if (sl_it == _service_levels_db.end() || sl_it->second.marked_for_deletion) {
```
Comment (avikivity) [inline service/qos/service_level_controller.hh:225]: Please fold into the commit that introduced the problem.

**Comment by avikivity** [comment]
Looks reasonable to me. Let me know if you want it merged. Please drop the RFC explanations from the cover letter.

File: service/qos/service_level_controller.hh (hunk 6 above)
Comment (psarna) [reply service/qos/service_level_controller.hh:225]: I did it that way to match the fix in enterprise (I generally tried to not touch the original series too much), but I can fold in here, no problem

**Comment by psarna** [comment]
It can be merged soon(ish), but first I need to put more work in #7913 in order to do proper testing of this one.

File: cql3/Cql.g (hunk 5 above)
Comment (psarna) [reply cql3/Cql.g:1892]: Looks like an oversight. Will fix

**Comment by psarna** [comment]
Rebased, green light for going through with this one. Even if something still needs to be changed in the future, it can be done in the form of follow-ups. I tested by basing #7913 on top of this one, because this series introduces a rather uninteresting mechanism by itself - a notion of service levels with 0 tweakable parameters.

**Comment by avikivity** [comment]
I see there are some LICENSE.PROPRIETARY leftovers there, please fix. I queued but immediately backed it out.

**Comment by psarna** [comment]
Repushed with proper open-source licensing.

**Comment by avikivity** [comment]
Requeued.

**Pull request merged**